plugins/inventory/iocage.py

```diff
@@ -0,0 +1,253 @@
+# -*- coding: utf-8 -*-
+
+# Copyright (c) 2024 Vladimir Botka <vbotka@gmail.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = '''
+    name: iocage
+    short_description: iocage inventory source
+    version_added: 10.2.0
+    author:
+        - Vladimir Botka (@vbotka)
+    requirements:
+        - iocage >= 1.8
+    description:
+        - Get inventory hosts from the iocage jail manager.
+        - Uses a configuration file as an inventory source, it must end
+          in C(.iocage.yml) or C(.iocage.yaml).
+    extends_documentation_fragment:
+        - constructed
+        - inventory_cache
+    options:
+        plugin:
+            description:
+              - The name of this plugin, it should always be set to
+                V(community.general.iocage) for this plugin to recognize
+                it as its own.
+            required: true
+            choices: ['community.general.iocage']
+            type: str
+        host:
+            description: The ip/name of the C(iocage) host.
+            type: str
+            default: localhost
+        user:
+            description:
+              - C(iocage) user.
+                It is expected that the O(user) is able to connect to the
+                O(host) and execute the command C(iocage list).
+                This option is not required if O(host) is V(localhost).
+            type: str
+        get_properties:
+            description:
+              - Get jails' properties.
+                Creates dictionary C(iocage_properties) for each added host.
+            type: boolean
+            default: false
+        env:
+            description: O(user)'s environment on O(host).
+            type: list
+            elements: str
+            default: []
+    notes:
+      - You might want to test the command C(ssh user@host iocage list -l)
+        on the controller before using this inventory plugin.
```

Comment:
I believe it is important to let the users know that this plugin does not allow specifying passwords nor encryption keys to perform the ssh operation on the remote node, and they should make use of `~/.ssh/config` to control those and other settings.
Also, I am thinking that some users might assume, mistakenly, that the ssh aspect of this will be handled by Ansible itself (as it uses ssh by default to reach remote targets).

TBH, I have mixed feelings about this usage of ssh, but I do not have a better way to do this. I cannot help thinking that if I were to use this, I might prefer making it in two steps: one ansible playbook to fetch the file from the remote, another one actually using it. As I wrote this, I recalled there was a case in `ignore-X.X.txt` files about a plugin that used a module: `iptables_state` action plugin _uses_ the `iptables_state` module. I have never fully understood this, nor have I spent time trying to, but it just _might_ provide you with a mechanism to let Ansible handle the ssh part for you. 

@felixfontein any insights on this?

Reply:
> ... I might prefer making it in two steps: one ansible playbook to fetch the file from the remote, another one actually using it.

If you prefer a playbook use the role [vbotka.freebsd_iocage](https://galaxy.ansible.com/ui/standalone/roles/vbotka/freebsd_iocage/).

Reply:
>  ... some users might assume, mistakenly, that the ssh aspect of this will be handled by Ansible 

I agree. Where would you like to put it? There are only the notes below atm.

```yaml
    notes:                                                                                                           
      - You might want to test the command C(ssh user@host iocage list -l)                                           
        on the controller before using this inventory plugin.                                                        
      - If you run this inventory plugin on V(localhost) C(ssh) is not used.                                         
        In this case, test the command C(iocage list -l).
        ...
```

Reply:
> ... it just _might_ provide you with a mechanism to let Ansible handle the ssh part for you.

Thank you for the effort. I'd like to keep it as simple as possible. By default, you don't need root to run *iocage* and get the data. There might be use cases where mandatory Ansible access would be insecure. Just like with any other VM manager. You get access to the VMs not to the manager.

Reply:
> And as I wrote it, I recalled there was a case in ignore-X.X.txt files about a plugin that used a module: iptables_state action plugin uses the iptables_state module. I have never fully understood this, nor have I spent time trying to, but it just might provide you with a mechanism to let Ansible handle the ssh part for you.

Action plugins are special because they are the controller version of modules, and users cannot distinguish between modules and action plugins (and usually action plugins come with modules, if only for documentation). So this is not related to inventory plugins at all :)

Reply:
> TBH, I have mixed feelings about this use of ssh, but I do not have a better way to do this. I cannot help thinking that if I were to use this, I might prefer making it in two steps: one ansible playbook to fetch the file from the remote, another one actually using it.

I agree. It would be nice to be able to stack plugins (connection plugins come to mind here, but also become plugins, and inventory plguins could use both) - but that isn't possible at the moment. I think I have seen at least one attempt of a plugin trying to load another plugin for this reason, but that's something I wouldn't recommend - at least for now. It might be worth trying this out in another collection (that's less stable than c.g) to see whether it's a direction worth pursuing...

Reply:
> > ... some users might assume, mistakenly, that the ssh aspect of this will be handled by Ansible
> 
> The parameter _user_ says:
> 
> ```yaml
>         user:
>             description:
>               - C(iocage) user.
>                 It is expected that the O(user) is able to connect to the
>                 O(host) and execute the command C(iocage list).
>                 This option is not required if O(host) is V(localhost).
> ```
> 
> Also,
> 
> ```yaml
>     notes:                                                                                                           
>       - You might want to test the command C(ssh user@host iocage list -l)                                           
>         on the controller before using this inventory plugin.                                                        
>       - If you run this inventory plugin on V(localhost) C(ssh) is not used.                                         
>         In this case, test the command C(iocage list -l).
>         ...
> ```

Not going to press the point, however I still think it is worth to mention that this ssh must work without interaction (e.g. keys without passphrase and/or other .ssh/config gimmicks). Again, it's your call.

Reply:
Thank you. I added `non-interactively` to the description item

```
        - By default, O(host) is V(localhost). If O(host) is not V(localhost) it
          is expected that the O(user) is able to connect to the O(host) with
          SSH non-interactively and execute the command C(iocage list).
```